subsystem/src/main/java/org/wildfly/extension/grpc/GrpcAttribute.java

```diff
@@ -0,0 +1,129 @@
+ *  limitations under the License.
+ */
+package org.wildfly.extension.grpc;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamWriter;
+
+import org.jboss.as.controller.AttributeDefinition;
+import org.jboss.as.controller.DefaultAttributeMarshaller;
+import org.jboss.as.controller.SimpleAttributeDefinition;
+import org.jboss.as.controller.SimpleAttributeDefinitionBuilder;
+import org.jboss.as.controller.operations.validation.ModelTypeValidator;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.ModelType;
+import org.jboss.dmr.Property;
+
+/**
+ * gRPC configuration attributes.
+ *
+ * @author <a href="mailto:rsigal@redhat.com">Ron Sigal</a>
+ */
+public abstract class GrpcAttribute {
+
+    public static final String GRPC_PARAMETER_GROUP = "gRPC";
+
+    public static final SimpleAttributeDefinition WILDFLY_GRPC_SERVER_PORT = new SimpleAttributeDefinitionBuilder(
+            GrpcConstants.WILDFLY_GRPC_SERVER_PORT, ModelType.INT)
+            .setRequired(false)
+            .setAllowExpression(true)
+            .setValidator(new ModelTypeValidator(ModelType.INT, false))
+            .setDefaultValue(new ModelNode(9555))
+            .setAttributeGroup(GRPC_PARAMETER_GROUP)
+            .build();
+
+    public static final SimpleAttributeDefinition WILDFLY_GRPC_SERVER_HOST = new SimpleAttributeDefinitionBuilder(
+            GrpcConstants.WILDFLY_GRPC_SERVER_HOST, ModelType.STRING)
+            .setRequired(false)
+            .setAllowExpression(true)
+            .setValidator(new ModelTypeValidator(ModelType.STRING, false))
+            .setDefaultValue(new ModelNode("0.0.0.0"))
+            .setAttributeGroup(GRPC_PARAMETER_GROUP)
+            .build();
+
+    public static final AttributeDefinition[] ATTRIBUTES = new AttributeDefinition[] {
+            WILDFLY_GRPC_SERVER_PORT,
+            WILDFLY_GRPC_SERVER_HOST,
+    };
+
+    public static final AttributeDefinition[] simpleAttributesArray = new AttributeDefinition[] {
+            WILDFLY_GRPC_SERVER_PORT,
+            WILDFLY_GRPC_SERVER_HOST,
+    };
+
+    public static final AttributeDefinition[] listAttributeArray = new AttributeDefinition[] {
+    };
+
+    public static final AttributeDefinition[] jndiAttributesArray = new AttributeDefinition[] {
+    };
+
+    public static final AttributeDefinition[] mapAttributeArray = new AttributeDefinition[] {
+    };
+
+    public static final Set<AttributeDefinition> SIMPLE_ATTRIBUTES = new HashSet<>(Arrays.asList(simpleAttributesArray));
+    public static final Set<AttributeDefinition> LIST_ATTRIBUTES = new HashSet<>(Arrays.asList(listAttributeArray));
+    public static final Set<AttributeDefinition> JNDI_ATTRIBUTES = new HashSet<>(Arrays.asList(jndiAttributesArray));
+    public static final Set<AttributeDefinition> MAP_ATTRIBUTES = new HashSet<>(Arrays.asList(mapAttributeArray));
```

Comment:
Upfront warning. I'm going to be nit-picky :)

I think these should be removed. There is not a strong reason to differentiate them.

Reply:
Removed